electrons/python/generic_electron_solver.py

```diff
@@ -0,0 +1,224 @@
+            )
+
+            von_neumann_expression = criterion_1 and criterion_2
+
+    def __post_init__(
+        self,
+    ):
+        # depending on the cluster/computer, the upper limit may be changed
+        if (self.no_xy * self.no_xy * self.no_z) > 1e8:
+            raise ValueError(
+                "Too many elements in the array: %i"
+                % (self.no_xy * self.no_xy * self.no_z_with_buffer)
+            )
+
+        self.von_neumann_expression()
+
+    @abstractmethod
+    def get_electron_density_after_beam(
+        self, positive_array: NDArray, negative_array: NDArray, time_step: int
+    ) -> Tuple[NDArray, NDArray, float]:
+        pass
+
+    def calculate(self):
+        positive_array = np.zeros((self.no_xy, self.no_xy, self.no_z_with_buffer))
+        negative_array = np.zeros((self.no_xy, self.no_xy, self.no_z_with_buffer))
+        positive_array_temp = np.zeros((self.no_xy, self.no_xy, self.no_z_with_buffer))
+        negative_array_temp = np.zeros((self.no_xy, self.no_xy, self.no_z_with_buffer))
+        no_recombined_charge_carriers = 0.0
+        no_initialised_charge_carriers = 0.0
+
+        """
+        Create an array which includes the number of track to be inserted for each time step
+        The tracks are distributed uniformly in time
+        """
+
+        positive_temp_entry = negative_temp_entry = recomb_temp = 0.0
+
+        f_steps_list = np.zeros(self.computation_time_steps)
+
+        szcz_pos = self.sz + self.cz * (self.cz + 1.0) / 2.0
+        szcz_neg = self.sz + self.cz * (self.cz - 1.0) / 2.0
+
+        sycy_pos = self.sy + self.cy * (self.cy + 1.0) / 2.0
+        sycy_neg = self.sy + self.cy * (self.cy - 1.0) / 2.0
+
+        sxcx_pos = self.sx + self.cx * (self.cx + 1.0) / 2.0
+        sxcx_neg = self.sx + self.cx * (self.cx - 1.0) / 2.0
+
+        cxyzsyz = (
+            1.0
+            - self.cx * self.cx
+            - self.cy * self.cy
+            - self.cz * self.cz
+            - 2.0 * (self.sx + self.sy + self.sz)
+        )
+
+        """
+        Start the simulation by evolving the distribution one step at a time
+        """
+
+        for time_step in range(self.computation_time_steps):
+
+            """
+            Refill the array with the electron density each time step
+            """
+            (
+                positive_array,
+                negative_array,
+                initialised_step,
+            ) = self.get_electron_density_after_beam(
+                positive_array,
+                negative_array,
+                time_step,
+            )
+
+            no_initialised_charge_carriers += initialised_step
+
+            # calculate the new densities and store them in temporary arrays
+            for i in range(1, self.no_xy - 1):
+                for j in range(1, self.no_xy - 1):
+                    for k in range(1, self.no_z_with_buffer - 1):
+                        # using the Lax-Wendroff scheme
```

Comment:
Maybe use numpy instead?

Reply:
Maybe use convolutions?

Reply:
Let's leave this as is - there isn't a good convolution for 3d spaces in numpy and the scipy implementation conflicts with numba.